sources/models/src/modeled_types/shared.rs

```diff
@@ -365,25 +366,36 @@ impl TryFrom<&str> for DNSDomain {
     type Error = error::Error;
 
     fn try_from(input: &str) -> Result<Self, Self::Error> {
-        ensure!(input.len() <= 253, error::InvalidDomainName { input });
-
-        let labels = input.split(".");
-        for (i, label) in labels.enumerate() {
-            ensure!(i < 127, error::InvalidDomainName { input });
-            ensure!(label.len() <= 63, error::InvalidDomainName { input });
-            let last = label.len() - 1;
-            for (j, c) in label.chars().enumerate() {
-                ensure!(
-                    (j != 0 && j != last && (c.is_alphanumeric() || c == '-'))
-                        || c.is_alphanumeric(),
-                    error::InvalidDomainName { input }
-                );
+        if input.starts_with('.') {
+            return error::InvalidDomainName {
+                input: input,
+                msg: "must not start with '.'",
             }
+            .fail();
         }
 
-        Ok(Self {
-            inner: input.to_string(),
-        })
+        match Host::parse(input) {
+            Err(e) => error::InvalidDomainName {
+                input: input,
+                msg: e.to_string(),
+            }
+            .fail(),
+            Ok(h) => match h {
+                Host::Ipv4(_) => error::InvalidDomainName {
+                    input: input,
+                    msg: "IP address is not a valid domain name",
+                }
+                .fail(),
+                Host::Ipv6(_) => error::InvalidDomainName {
+                    input: input,
+                    msg: "IP address is not a valid domain name",
+                }
+                .fail(),
+                Host::Domain(_) => Ok(Self {
+                    inner: input.to_string(),
+                }),
+            },
+        }
```

Comment:
```suggestion
        let host = Host::parse(input).map_err(|e| error::InvalidDomainName {
            input: input,
            msg: e.to_string(),
        })?;
        match host {
            Host::Ipv4(_) | Host::Ipv6(_) => error::InvalidDomainName {
                input: input,
                msg: "IP address is not a valid domain name",
            }
            .fail(),
            Host::Domain(_) => Ok(Self {
                inner: input.to_string(),
            }),
        }
```
A little simplification; I haven't tested this, but I think it'll work.

Reply:
Thanks. Like I mentioned, I'm new to rust. This helps a lot. I don't have any sense of the best way to do things yet.

Reply:
No problem!  This is a little more idiomatic, I think.  I hope you're enjoying Rust - it's been great for Bottlerocket.

We can merge the PR as soon as you're happy with these last changes.

Reply:
I hate to bug you with language questions. But I can't seem to find a good way to deal with this:

```rust
        let host = Host::parse(input).map_err(|e| error::InvalidDomainName {
            input: input,
            msg: e.to_string(),
        })?;
```

The `?` needs to use the From trait to convert from a `modeled_types::error::InvalidDomainName` to a `modeled_types::error::Error`, but that trait isn't implemented. What's the best way to handle that?

Reply:
No problem, I'm sorry I gave an incomplete suggestion!  I've updated it above, and it compiles now.

The problem you ran into is related to how `snafu`, the error handling library we use, works.  The details are [here](https://docs.rs/snafu/0.6.8/snafu/guide/the_macro/index.html), but the short version is: snafu provides a nice interface to library developers using a procedural macro that adds to the code at compile time.  The library developer can make a nice enum of their error cases, and they don't have to define everything about error handling that's actually needed.  snafu takes that and generates structs for each of the error variants, which are what's actually used behind the scenes, and makes error construction easier.

My mistake was suggesting `map_err` and creating the struct form (`modeled_types::error::InvalidDomain`) inside, when the return type of the function needs the enum form (`modeled_types::error::Error`).  Rust helpfully tried to convert from one to the other with `From` as part of using `?` for an early return, but it doesn't know how.  What I did in my revised suggestion is to convert from the struct form to the enum form with `.fail()` so we can return it.  (This also means I needed to use `or_else` instead of `map_err`; they're identical except that `or_else` takes a full `Result` and not just the `Err` part inside a result, which we need because `fail()` returns a full `Result`, for reasons related to ease of use in more common scenarios.)

Hope this helps, and let me know if you have any other trouble.

Reply:
Thanks for the explanation! This has been very helpful for me.